rfd/0033-sni-alpn-teleport-proxy-routing.md

```diff
@@ -0,0 +1,221 @@
+
+#### Setting SNI value for kubectl CLI:
+
+`tsh kube login` command generates a local kubeconfig.yaml file used during accessing teleport proxy by kubectl CLI. Additional to the current configuration the  tls-server-name  failed will be added with appropriate SNI value:
+
+```yaml
+apiVersion: v1
+clusters:
+- cluster:
+  certificate-authority-data: ...
+  server: public_proxy_ip:443
+  tls-server-name: kube.proxy.example.teleport.sh
+  name: proxy.example.teleport.sh
+```
+
+
+### Local teleport proxy
+
+To handle connections established to teleport proxy by external clients like mysql, psql CLIs where setting custom ALPN value is not possible we will run local teleport proxy. Traffic sent from external teleport clients will be forwarded through the local proxy where the proxy will be responsible for wrapping incoming connections in TLS protocol, setting appropriate ALPN protocol and establishing a connection to the remote teleport proxy.
+
+For the tsh ssh and tsh scp commands scope forwarding traffic through local proxy seems to be superfluous. SSH traffic can be easily wrapped in TLS with custom teleport-ssh-proxy ALPN protocol inside tsh ssh handler after detecting that the teleport proxy_listener_addr listener was enabled but we should consider adding support for ssh proxy in order to enable external SSH and SCP clients like FileZilla, WinSCP and Putty.
+
+Question:
+- Do we want to support Unix sockets in local proxy ?
+
+
+#### OpenSSH ProxyCommand
+In order to support connection from OpenSSH clients where traffic send to cluster proxy is wrapped in TLS protocol the tsh binary will provide a new `tsh proxy ssh` subcommand that can be injected to ProxyCommand like `ssh -o "ForwardAgent yes" -o "ProxyCommand tsh proxy ssh" alice@nodeone.example.com"`
+The `tsh proxy ssh` command will be responsible for establishing a TLS connection to cluster proxy with ALPN: teleport-reverse-tunnel protocol, forwarding ssh-agent and proxying traffic between an openssh client and cluster proxy.
+
+
+### Reverse tunnel
+Currently, reverse tunnel connection uses SSH protocol, and a separate cluster proxy port.  Reverse tunnels connection established from proxy client will be  wrapped in TLS  protocol with `teleport-reverse-tunnel` SNI where cluster proxy TLS listener will be
+responsible for TLS termination and forwarding incoming SSH traffic to the `reversetunnel` teleport service where further remains unchanged.
+
+Internal reverse tunnel dialer() will be aligned and an established reverse tunnel cluster proxy ssh connection will be wrapped in TLS
+
+### CLI UX
+
+#### kubectl CLI
+kubectl CLI reads the cluster config from the `KUBECONFIG` file thus, UX remains unchanged. The proper SNI value will be set automatically based on `tls-server-name` field that will be set by `tsh kube login` command.
+
+#### APP curl access
+The app access UX will remain unchanged, `curl` already sets `ALPN` protocols to `h2,http/1.1` and SNI to the destination URL host.
```

Comment:
curl is just one example, API access should work with anything that can make http requests with client certs - but I guess it will work automatically since the traffic will just be detected as "http"?